File: CppGen/CppGen/Object.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace CppGen
{
	// A GameMaker object or a struct in the code.
	// For objects, the Create Event is the constructor and Destroy Event the destructor.
	// For structs, the declaration is the contructor, with no destructor. Static functions are added into InstanceFunctions.
	public class Object
	{
		public string Name;
		public Function CreateFunction = null;
		public Function DestroyFunction = null;
		public Function Constructor = null;
		public Function Destructor = null;
		public Dictionary<string, Function> InstanceFunctions = new Dictionary<string, Function>();
		public Dictionary<string, Variable> InstanceVars = new Dictionary<string, Variable>();
		public bool IsStruct = false;
		
		private bool ImplWritten = false;

		public Object(string name, bool isStruct)
		{
			Name = name;
			IsStruct = isStruct;
		}

		public Object(string dir) : this("", false)
		{
			DirectoryInfo dirInfo = new DirectoryInfo(dir);
			Name = dirInfo.Name;

			// Parse selected events and their script
			string[] objectEventFiles = Directory.GetFiles(dir, "*.gml");
			foreach (string eventFile in objectEventFiles)
			{
				FileInfo eventFileInfo = new FileInfo(eventFile);
				string gml = File.ReadAllText(eventFile);
				string funcName = gml.Replace("()", "").Replace(";", "").Trim();

				if (!Program.Functions.ContainsKey(funcName))
					continue;
				Function func = Program.Functions[funcName];

				if (eventFileInfo.Name == "Create_0.gml")
					CreateFunction = func;
				else if (eventFileInfo.Name == "Destroy_0.gml")
					DestroyFunction = func;
				else if (dirInfo.Name == "app")
				{
					switch (eventFileInfo.Name)
					{
						case "Draw_0.gml": Program.AppDrawFunction = func; break;
						case "Other_3.gml": Program.AppGameEndFunction = func; break;
						case "Other_62.gml": Program.AppHttpFunction = func; break;
						case "Step_0.gml": Program.AppStepFunction = func; break;
						default:
						{
							Console.WriteLine("WARNING: Unsupported event {0} in {1}, it will be ignored.", eventFileInfo.Name, "app");
							break;
						}
					}
				}
				else
					Console.WriteLine("WARNING: Unsupported event {0} in {1}, it will be ignored." + eventFileInfo.Name, dirInfo.Name);
			}
		}

		// Sets the constructor of the object.
		public void SetConstructor(Function func)
		{
			Constructor = func;
			func.StructObject = this;
			func.IsConstructor = true;
			func.CppLinesBegin.Add("InitMembers();");
		}

		// Sets the destructor of the object.
		public void SetDestructor(Function func)
		{
			Destructor = func;
			func.StructObject = this;
			func.IsDestructor = true;
		}

		// Writes the C++ header of the object.
		public void WriteCppHeader()
		{
			CodeWriter.WriteLine("struct " + Name + ": Object");
			CodeWriter.WriteLine("{", 1);

			// Variables
			foreach (Variable var in InstanceVars.Values)
				if (var.Name != "id" && var.Name != "subAssetId")
					CodeWriter.WriteLine(var.Type.ToCpp() + " " + CodeObject.NameToCpp(var.Name) + ";");

			if (Constructor != null || Destructor != null || InstanceFunctions.Values.Count > 0)
				CodeWriter.WriteLine();

			// Constructor header
			if (Constructor != null)
			{
				CodeWriter.Write(Name + "(");
				if (Constructor.Args != null)	
					Constructor.Args.WriteCpp(new ResolveScope("global"), DeclarationList.WriteFormat.ArgsHeader);
				CodeWriter.WriteLine(");");
			}

			// Destructor header
			if (Destructor != null)
				CodeWriter.WriteLine("~" + Name + "();");

			CodeWriter.WriteLine("void InitMembers() override;");

			// Instance functions
			foreach (Function func in InstanceFunctions.Values)
				func.WriteCppHeader();

			CodeWriter.WriteLine("};", -1);
		}

		// Writes the C++ implementation of the object. Returns whether anything was written.
		public bool WriteCppImplementation()
		{
			if (ImplWritten)
				return false;
			ImplWritten = true;

			// Constructor
			if (Constructor != null)
				Constructor.WriteCppImplementation();

			// Destructor
			if (Destructor != null)
				Destructor.WriteCppImplementation();

			foreach (Function func in InstanceFunctions.Values)
				func.WriteCppImplementation();
			return true;
		}

		// Writes the commands for storing the memory locations of each object member and functions.
		public void WriteInitMembers()
		{
			CodeWriter.WriteLine("void " + Name + "::InitMembers()");
			CodeWriter.WriteLine("{", 1);

			// Define instance functions for use with the idFunc macro
			if (InstanceFunctions.Count > 0)
			{
				foreach (Function func in InstanceFunctions.Values)
					CodeWriter.WriteLine("DefineObjectFunction(ID_" + Name + ", M_" + func.Name + ", " + func.ToExecuteCpp() + ")");
				CodeWriter.WriteLine();
			}

			CodeWriter.WriteLine("if (memberMap.contains(ID_" + Name + "))", 1);
			CodeWriter.WriteLine("return;");
			CodeWriter.WriteLine("", -1);

			// Define members
			foreach (Variable var in InstanceVars.Values)
				CodeWriter.WriteLine("DefineObjectMember(ID_" + Name + ", M_" + CodeObject.NameToCpp(var.Name) + ", " + CodeObject.NameToCpp(var.Name) + ", " + var.Type.ToCppEnum() + ");");
			
			CodeWriter.WriteLine("}", -1);
			CodeWriter.WriteLine();
		}

		// Returns a string containing the functions and variables of the object.
		public string ToDebugString(string tabs = "")
		{
			string str = tabs + Name + ":\n";
			tabs += "\t";
			if (Constructor != null)
				str += tabs + "Constructor: " + Constructor.Name + "\n";
			if (Destructor != null)
				str += tabs + "Destructor: " + Destructor.Name + "\n";
			str += tabs + "LocalFunctions:\n";
			foreach (Function func in InstanceFunctions.Values)
				str += func.ToDebugString(tabs + "\t");
			str += tabs + "Variables:\n";
			tabs += "\t";
			foreach (Variable var in InstanceVars.Values)
				str += tabs + var.Type.ToCpp() + " " + var.Name + "\n" + var.Type.GetAssignmentsString(tabs + "\t");
			return str;
		}
	}
}

File: CppGen/Sources/Object.cpp
#include "CppGen.hpp"

namespace CppGen
{
Object::Object(StringId name, bool isStruct)
{
	this->name = name;
	this->isStruct = isStruct;
}

Object::Object(String dir)
{
	DirectoryInfo dirInfo = DirectoryInfo(dir);
	this->name = dirInfo.name;

	// Parse selected events and their script
	List<String> objectEventFiles = Directory::getFiles(dir, "*.gml");
	for (String eventFile : objectEventFiles)
	{
		FileInfo eventFileInfo = FileInfo(eventFile);
		String gml = File::readAllText(eventFile);
		String funcName = gml.replace("()", "").replace(";", "").trim();

		if (!Program::functions.containsKey(funcName))
			continue;
		Function* func = Program::functions[funcName];

		if (eventFileInfo.name == "Create_0.gml")
			this->createFunction = func;
		else if (eventFileInfo.name == "Destroy_0.gml")
			this->destroyFunction = func;
		else if (dirInfo.name == "app")
		{
			if (eventFileInfo.name == "Draw_0.gml")
				Program::appDrawFunction = func;
			else if (eventFileInfo.name == "Other_3.gml")
				Program::appGameEndFunction = func;
			else if (eventFileInfo.name == "Other_62.gml")
				Program::appHttpFunction = func;
			else if (eventFileInfo.name == "Step_0.gml")
				Program::appStepFunction = func;
			else
				Console::writeLine("WARNING: Unsupported event {0} in {1}, it will be ignored.", eventFileInfo.name, "app");
		}
		else
			Console::writeLine("WARNING: Unsupported event {0} in {1}, it will be ignored." + eventFileInfo.name, dirInfo.name);
	}
}

void Object::setConstructor(Function* func)
{
	this->constructor = func;
	func->structObject = this;
	func->isConstructor = true;
	func->cppLinesBegin.add("InitMembers();");
}

void Object::setDestructor(Function* func)
{
	this->destructor = func;
	func->structObject = this;
	func->isDestructor = true;
}

void Object::writeCppHeader()
{
	CodeWriter::writeLine("struct " + String(this->name) + ": Object");
	CodeWriter::writeLine("{", 1);

	// Variables
	for (Variable* var : this->instanceVars.values)
		if (var->name != STR(id) && var->name != STR(subAssetId))
			CodeWriter::writeLine(var->type->toCpp() + " " + CodeObject::nameToCpp(var->name) + ";");

	if (this->constructor != nullptr || this->destructor != nullptr || this->instanceFunctions.values.size() > 0)
		CodeWriter::writeLine();

	// Constructor header
	if (this->constructor != nullptr)
	{
		CodeWriter::write(String(this->name) + "(");
		if (this->constructor->args != nullptr)
			this->constructor->args->writeCpp(ResolveScope(STR(global)), DeclarationList::WriteFormat::ArgsHeader);
		CodeWriter::writeLine(");");
	}

	// Destructor header
	if (this->destructor != nullptr)
		CodeWriter::writeLine("~" + String(this->name) + "();");

	CodeWriter::writeLine("void InitMembers() override;");

	// Instance functions
	for (Function* func : this->instanceFunctions.values)
		func->writeCppHeader();

	CodeWriter::writeLine("};", -1);
}

bool Object::writeCppImplementation()
{
	if (this->implWritten)
		return false;
	this->implWritten = true;

	// Constructor
	if (this->constructor != nullptr)
		this->constructor->writeCppImplementation();

	// Destructor
	if (this->destructor != nullptr)
		this->destructor->writeCppImplementation();

	for (Function* func : this->instanceFunctions.values)
		func->writeCppImplementation();
	return true;
}

void Object::writeInitMembers()
{
	CodeWriter::writeLine("void " + String(this->name) + "::InitMembers()");
	CodeWriter::writeLine("{", 1);

	// Define instance functions for use with the idFunc macro
	if (this->instanceFunctions.size() > 0)
	{
		for (Function* func : this->instanceFunctions.values)
			CodeWriter::writeLine("DefineObjectFunction(ID_" + String(this->name) + ", M_" + String(func->name) + ", " + String(func->toExecuteCpp()) + ")");
		CodeWriter::writeLine();
	}

	CodeWriter::writeLine("if (memberMap.contains(ID_" + String(this->name) + "))", 1);
	CodeWriter::writeLine("return;");
	CodeWriter::writeLine("", -1);

	// Define members
	for (Variable* var : this->instanceVars.values)
		CodeWriter::writeLine("DefineObjectMember(ID_" + String(this->name) + ", M_" + CodeObject::nameToCpp(var->name) + ", " + CodeObject::nameToCpp(var->name) + ", " + var->type->toCppEnum() + ");");

	CodeWriter::writeLine("}", -1);
	CodeWriter::writeLine();
}

}
